plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/explorer/OpenShiftExplorerContentProvider.java

```diff
@@ -75,6 +65,58 @@ protected void handleConnectionChanged(IConnection connection, String property,
 			super.handleConnectionChanged(connection, property, oldValue, newValue);
 		}
 	}
+	private void handleAddResource(IResource resource) {
+		switch(resource.getKind()) {
+		case ResourceKind.SERVICE:
+			IService service = (IService) resource;
+			Deployment deployment = DeploymentResourceMapper.getDeployment(service);
+			deploymentByService.put(service, deployment);
+			addChildrenToViewer(resource.getProject(), deployment);
+			break;
+		case ResourceKind.POD:
+			IPod pod = (IPod)resource;
+			Map<String, String> labels = pod.getLabels();
+			for (IService depService : deploymentByService.keySet()) {
+				Map<String, String> selector = depService.getSelector();
+				if(selectorsOverlap(selector, labels)) {
+					Deployment deploymentForPod = deploymentByService.get(depService);
+					deploymentForPod.add(pod);
+					addChildrenToViewer(depService, deploymentForPod);
+				}
+			}
+		default:
+		}
+	}
+	private void handleRemoveResource(IResource resource) {
+		switch(resource.getKind()) {
+		case ResourceKind.SERVICE:
+			IService service = (IService)resource;
+			deploymentByService.remove(service);
+			removeChildrenFromViewer(resource.getProject(), deploymentByService.get(service));
+			break;
+		case ResourceKind.POD:
+			IPod pod = (IPod)resource;
+			for (IService deployedService : deploymentByService.keySet()) {
+				if(selectorsOverlap(deployedService.getSelector(), pod.getLabels())){
+					deploymentByService.get(deployedService).remove(pod);
+					removeChildrenFromViewer(deploymentByService, pod);
```

Comment:
I would ignore this 'refresh' logic for now since its broken in this pull.  It needs to be investigated more as a follow on task.  Your comments may be valid